daisy/serial_output_watcher.go

```diff
@@ -0,0 +1,165 @@
+		project:       project,
+		zone:          zone,
+	}
+}
+
+type serialOutputWatcher struct {
+	subscriptions map[instanceAndPort][]watch
+	client        serialOutputClient
+	project, zone string
+}
+
+// serialOutputWatcher assumes that 404s are returned when an instance doesn't
+// exist. The value `deletionNum404s` determines how many subsequent HTTP 404 responses to
+// see prior to deciding the instance has been deleted (assuming the
+// instance was seen earlier).
+//
+// Although a value of 1 is theoretically sufficient, a higher value allows withstanding
+// unexpected intermittent failures.
+//
+// At the time of writing, the following HTTP error codes were seen:
+//  - Prior to creating instance: 404
+//  - Once deletion finished: 404
+//  - Instance stopped, but not deleted: 400
+//  - Serial output successfully returned: 200
+const deletionNum404s = 3
+
+type instanceAndPort struct {
+	instance   string
+	portNumber int64
+}
+
+type watch struct {
+	channel         chan<- string
+	pollingInterval time.Duration
+}
+
+func (s *serialOutputWatcher) Watch(
+	instanceName string, port int64, c chan<- string, interval time.Duration) {
+	sp := instanceAndPort{instanceName, port}
+	s.subscriptions[sp] = append(s.subscriptions[sp], watch{c, interval})
+}
+
+func (s *serialOutputWatcher) start(instanceName string) {
+	for serialPort := range s.subscriptions {
+		if serialPort.instance != instanceName {
+			continue
+		}
+		callbacks := s.subscriptions[serialPort]
+		if len(callbacks) == 0 {
+			panic(fmt.Sprintf("Start called without subscribers for %v", serialPort))
+		}
+		var subscribers []chan<- string
+		pollingFrequency := callbacks[0].pollingInterval
+		for _, cb := range callbacks {
+			subscribers = append(subscribers, cb.channel)
+			if cb.pollingInterval < pollingFrequency {
+				cb.pollingInterval = pollingFrequency
+			}
+		}
+		watchOnePort(s.client, pollingFrequency, s.project, s.zone, serialPort.instance, serialPort.portNumber, subscribers)
+	}
+}
+
+// run starts a goroutine that polls for an instance's serial output. Output is written to the
+// subscribing channels. Channels are closed when the instance is deleted.
+func watchOnePort(client serialOutputClient, pollingFrequency time.Duration,
+	project, zone, instanceName string, port int64, subscribers []chan<- string) {
+	go func() {
+		seenInstance := false
+		consecutive404s := 0
+		offset := int64(0)
+		ticker := time.NewTicker(pollingFrequency)
+		for {
+			resp, err := client.GetSerialPortOutput(project, zone, instanceName, port, offset)
+			var httpCode int
+			if err != nil {
+				realErr := err.(*googleapi.Error)
+				httpCode = realErr.Code
+			} else {
+				offset = resp.Next
+				httpCode = resp.HTTPStatusCode
+				seenInstance = true
+				if len(resp.Contents) > 0 {
+					for _, c := range subscribers {
+						select {
+						case c <- resp.Contents:
+						default:
+						}
+					}
+				}
+			}
+			if seenInstance && httpCode == http.StatusNotFound {
```

Comment:
Does this mean 404 is ignored until instance is created to avoid bailing on reading serial port output before instance is finished being created?